File: Sources/DebridStreamer/Services/Indexers/IndexerFactory.swift
import Foundation

enum IndexerFactory {
    static func buildIndexers(from configs: [IndexerConfig]) -> [TorrentIndexer] {
        var result: [TorrentIndexer] = []

        let builtInEnabled = configs.first(where: { $0.type == .builtIn })?.isActive ?? true
        if builtInEnabled {
            result.append(APIBayIndexer())
            result.append(YTSIndexer())
            result.append(EZTVIndexer())
        }

        let activeExternal = configs
            .filter { $0.isActive && $0.type != .builtIn }
            .sorted { $0.priority < $1.priority }

        for config in activeExternal {
            guard let indexer = makeExternalIndexer(from: config) else { continue }
            result.append(indexer)
        }

        return result
    }

    static func testConnection(config: IndexerConfig, session: URLSession = .shared) async -> Bool {
        guard config.type != .builtIn else {
            // Built-in indexers have no user-configured endpoint to validate.
            return true
        }

        if config.type == .stremioAddon {
            return await testStremioAddon(config: config, session: session)
        }

        // Validate the HTTP layer and require a positive signal (2xx AND a
        // parseable Torznab/RSS feed without an error envelope), rather than
        // merely "the request did not throw". A wrong base URL, bad path, or
        // bad API key typically yields a non-2xx status or an error envelope.
        guard let request = makeTorznabProbeRequest(from: config) else {
            return false
        }

        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse,
                  (200...299).contains(http.statusCode) else {
                return false
            }
            return isPositiveTorznabResponse(data)
        } catch {
            return false
        }
    }

    /// Validates a Stremio addon by fetching its `manifest.json`. A valid addon
    /// returns a 2xx JSON manifest carrying at least an `id` field.
    private static func testStremioAddon(config: IndexerConfig, session: URLSession) async -> Bool {
        var base = config.baseURL.trimmingCharacters(in: .whitespacesAndNewlines)
        while base.hasSuffix("/") { base.removeLast() }
        // Accept either a bare base URL or a full `.../manifest.json`.
        let manifestURLString = base.hasSuffix("manifest.json") ? base : "\(base)/manifest.json"
        guard let url = URL(string: manifestURLString) else { return false }

        var request = URLRequest(url: url)
        request.timeoutInterval = 20
        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse,
                  (200...299).contains(http.statusCode) else {
                return false
            }
            guard let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                return false
            }
            return object["id"] != nil || object["resources"] != nil
        } catch {
            return false
        }
    }

    /// Builds a minimal Torznab search probe request for `testConnection`.
    private static func makeTorznabProbeRequest(from config: IndexerConfig) -> URLRequest? {
        guard var components = URLComponents(string: config.baseURL) else { return nil }
        if !config.endpointPath.isEmpty {
            let current = components.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
            let append = config.endpointPath.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
            if current.isEmpty {
                components.path = "/\(append)"
            } else if append.isEmpty {
                components.path = "/\(current)"
            } else {
                components.path = "/\(current)/\(append)"
            }
        }

        let sendAPIKeyAsHeader = config.providerSubtype == .prowlarr
        var queryItems: [URLQueryItem] = [
            URLQueryItem(name: "t", value: "search"),
            URLQueryItem(name: "q", value: "test")
        ]
        if let categoryFilter = config.categoryFilter, !categoryFilter.isEmpty {
            queryItems.append(URLQueryItem(name: "cat", value: categoryFilter))
        }
        if let apiKey = config.apiKey, !apiKey.isEmpty, !sendAPIKeyAsHeader {
            queryItems.append(URLQueryItem(name: "apikey", value: apiKey))
        }
        components.queryItems = queryItems

        guard let url = components.url else { return nil }
        var request = URLRequest(url: url)
        request.timeoutInterval = 20
        if let apiKey = config.apiKey, !apiKey.isEmpty, sendAPIKeyAsHeader {
            request.setValue(apiKey, forHTTPHeaderField: "X-Api-Key")
        }
        return request
    }

    /// Returns true only when the body looks like a valid Torznab/RSS feed and
    /// is not a Torznab error envelope (e.g. `<error code="100" .../>`).
    private static func isPositiveTorznabResponse(_ data: Data) -> Bool {
        guard let body = String(data: data, encoding: .utf8) else { return false }
        let lower = body.lowercased()
        // A Torznab error envelope is a negative signal even with HTTP 200.
        if lower.contains("<error") {
            return false
        }
        // Require a recognizable feed root; a valid empty feed still passes.
        return lower.contains("<rss") || lower.contains("<?xml")
    }

    private static func makeExternalIndexer(from config: IndexerConfig, session: URLSession = .shared) -> (any TorrentIndexer)? {
        switch config.type {
        case .jackett, .prowlarr, .torznab, .zilean:
            let displayName = config.displayName?.trimmingCharacters(in: .whitespacesAndNewlines)
            let name = (displayName?.isEmpty == false) ? displayName! : config.type.displayName
            let sendAPIKeyAsHeader = config.providerSubtype == .prowlarr
            return TorznabIndexer(
                name: name,
                baseURL: config.baseURL,
                endpointPath: config.endpointPath,
                apiKey: config.apiKey,
                categoryFilter: config.categoryFilter,
                sendAPIKeyAsHeader: sendAPIKeyAsHeader,
                session: session
            )
        case .stremioAddon:
            let baseURL = config.baseURL.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !baseURL.isEmpty, URL(string: baseURL) != nil else { return nil }
            let displayName = config.displayName?.trimmingCharacters(in: .whitespacesAndNewlines)
            let name = (displayName?.isEmpty == false) ? displayName! : config.type.displayName
            return StremioAddonIndexer(name: name, baseURL: baseURL)
        case .builtIn:
            return nil
        }
    }
}

File: web/src/services/indexers/IndexerFactory.ts
// Port of Sources/DebridStreamer/Services/Indexers/IndexerFactory.swift.
//
// Builds the active indexer set from a list of IndexerConfigs (built-in
// APIBay/YTS/EZTV when enabled, plus priority-sorted external Torznab indexers)
// and validates an external endpoint via `testConnection` (2xx + a positive
// Torznab/RSS feed without an `<error>` envelope).

import { APIBayIndexer } from "./APIBayIndexer";
import { EZTVIndexer } from "./EZTVIndexer";
import { StremioAddonIndexer } from "./StremioAddonIndexer";
import {
  defaultFetchImpl,
  type FetchImpl,
  type IndexerConfig,
  ProviderSubtype,
  type TorrentIndexer,
} from "./types";
import { TorznabIndexer } from "./TorznabIndexer";
import { YTSIndexer } from "./YTSIndexer";

export const IndexerFactory = {
  /** Builds the active indexer list. Built-in scrapers come first (enabled
   * unless a `built_in` config is explicitly inactive), then active external
   * indexers sorted by ascending priority. Mirrors `buildIndexers`. */
  buildIndexers(
    configs: IndexerConfig[],
    fetchImpl: FetchImpl = defaultFetchImpl,
  ): TorrentIndexer[] {
    const result: TorrentIndexer[] = [];

    const builtInConfig = configs.find((c) => c.type === "built_in");
    const builtInEnabled = builtInConfig?.isActive ?? true;
    if (builtInEnabled) {
      result.push(new APIBayIndexer(fetchImpl));
      result.push(new YTSIndexer(fetchImpl));
      result.push(new EZTVIndexer(fetchImpl));
    }

    const activeExternal = configs
      .filter((c) => c.isActive && c.type !== "built_in")
      .sort((a, b) => a.priority - b.priority);

    for (const config of activeExternal) {
      const indexer = makeExternalIndexer(config, fetchImpl);
      if (indexer == null) continue;
      result.push(indexer);
    }

    return result;
  },

  /** Validates an external indexer endpoint. Built-in always passes; otherwise
   * requires 2xx AND a positive Torznab/RSS feed (no `<error>` envelope).
   * Mirrors `testConnection`. */
  async testConnection(
    config: IndexerConfig,
    fetchImpl: FetchImpl = defaultFetchImpl,
  ): Promise<boolean> {
    if (config.type === "built_in") {
      // Built-in indexers have no user-configured endpoint to validate.
      return true;
    }
    if (config.type === "stremio_addon") {
      return testStremioAddonConnection(config, fetchImpl);
    }

    const probe = makeTorznabProbeRequest(config);
    if (probe == null) return false;

    try {
      const response = await fetchImpl(
        probe.url,
        probe.headers ? { headers: probe.headers } : undefined,
      );
      if (!(response.status >= 200 && response.status <= 299)) {
        return false;
      }
      return isPositiveTorznabResponse(await response.text());
    } catch {
      return false;
    }
  },
} as const;

/** Validates a Stremio addon by fetching and parsing `manifest.json`. A valid
 * addon is 2xx with valid JSON containing either an `id` or `resources`
 * field, matching the Swift `testStremioAddon` behavior. */
async function testStremioAddonConnection(
  config: IndexerConfig,
  fetchImpl: FetchImpl,
): Promise<boolean> {
  const trimmedBase = config.baseURL.trim();
  if (trimmedBase.length === 0) return false;

  let manifestBase = trimmedBase.replace(/\/+$/, "");
  if (/\/manifest\.json$/i.test(manifestBase)) {
    manifestBase = manifestBase.replace(/\/manifest\.json$/i, "");
  }
  const target = `${manifestBase}/manifest.json`;

  try {
    const response = await fetchImpl(target);
    if (!(response.status >= 200 && response.status <= 299)) {
      return false;
    }

    const parsed = JSON.parse(await response.text());
    if (parsed == null || typeof parsed !== "object" || Array.isArray(parsed)) {
      return false;
    }

    return (
      Object.hasOwn(parsed as Record<string, unknown>, "id") ||
      Object.hasOwn(parsed as Record<string, unknown>, "resources")
    );
  } catch {
    return false;
  }
}

/** Builds a minimal Torznab search probe (url + optional headers).
 * Mirrors `makeTorznabProbeRequest`. */
function makeTorznabProbeRequest(
  config: IndexerConfig,
): { url: string; headers: Record<string, string> | null } | null {
  const trimmedBase = config.baseURL.trim();
  if (trimmedBase.length === 0) {
    return null;
  }
  const trimmedEndpointPath = config.endpointPath.trim();

  let url: URL;
  try {
    url = new URL(trimmedBase);
  } catch {
    return null;
  }

  if (trimmedEndpointPath.length > 0) {
    const current = trimSlashes(url.pathname);
    const append = trimSlashes(trimmedEndpointPath);
    if (current.length === 0) {
      url.pathname = `/${append}`;
    } else if (append.length === 0) {
      url.pathname = `/${current}`;
    } else {
      url.pathname = `/${current}/${append}`;
    }
  }

  const sendAPIKeyAsHeader = config.providerSubtype === ProviderSubtype.prowlarr;
  url.searchParams.append("t", "search");
  url.searchParams.append("q", "test");
  if (config.categoryFilter != null && config.categoryFilter.length > 0) {
    url.searchParams.append("cat", config.categoryFilter);
  }
  if (
    config.apiKey != null &&
    config.apiKey.length > 0 &&
    !sendAPIKeyAsHeader
  ) {
    url.searchParams.append("apikey", config.apiKey);
  }

  let headers: Record<string, string> | null = null;
  if (
    config.apiKey != null &&
    config.apiKey.length > 0 &&
    sendAPIKeyAsHeader
  ) {
    headers = { "X-Api-Key": config.apiKey };
  }

  return { url: url.toString(), headers };
}

/** True only when the body looks like a valid Torznab/RSS feed and is not a
 * Torznab error envelope (`<error code=... />`). Mirrors
 * `isPositiveTorznabResponse`. */
function isPositiveTorznabResponse(body: string): boolean {
  const lower = body.toLowerCase();
  // A Torznab error envelope is a negative signal even with HTTP 200.
  if (lower.includes("<error")) {
    return false;
  }
  // Require a recognizable feed root; a valid empty feed still passes.
  return lower.includes("<rss") || lower.includes("<?xml");
}

/** Builds an external (Torznab-family) indexer from a config, or null for
 * built-in. Mirrors `makeExternalIndexer`. */
function makeExternalIndexer(
  config: IndexerConfig,
  fetchImpl: FetchImpl,
): TorrentIndexer | null {
  if (config.type === "built_in") {
    return null;
  }

  switch (config.type) {
    case "jackett":
    case "prowlarr":
    case "torznab":
    case "zilean": {
      const trimmedBase = config.baseURL.trim();
      if (trimmedBase.length === 0) return null;
      const displayName = config.displayName?.trim();
      const name =
        displayName != null && displayName.length > 0
          ? displayName
          : indexerTypeDisplayName(config.type);
      const trimmedEndpointPath = config.endpointPath.trim();
      const sendAPIKeyAsHeader =
        config.providerSubtype === ProviderSubtype.prowlarr;
      return new TorznabIndexer({
        name,
        baseURL: trimmedBase,
        endpointPath: trimmedEndpointPath,
        apiKey: config.apiKey,
        categoryFilter: config.categoryFilter,
        sendAPIKeyAsHeader,
        fetchImpl,
      });
    }
    case "stremio_addon": {
      const baseURL = config.baseURL.trim();
      // Mirror the Swift factory: skip an empty / unparseable base URL.
      if (baseURL.length === 0) return null;
      try {
        // eslint-disable-next-line no-new
        new URL(baseURL);
      } catch {
        return null;
      }
      const displayName = config.displayName?.trim();
      const name =
        displayName != null && displayName.length > 0
          ? displayName
          : indexerTypeDisplayName(config.type);
      return new StremioAddonIndexer(name, baseURL, fetchImpl);
    }
    default:
      // Keep unknown types resilient and observably labeled even if a new backend
      // value slips through before a proper case exists.
      indexerTypeDisplayName(config.type);
      return null;
  }
}

function indexerTypeDisplayName(type: string): string {
  switch (type) {
    case "jackett":
      return "Jackett";
    case "prowlarr":
      return "Prowlarr";
    case "torznab":
      return "Torznab";
    case "zilean":
      return "Zilean";
    case "stremio_addon":
      return "Stremio Addon";
  }
  return `Unknown (${type})`;
}

function trimSlashes(s: string): string {
  return s.replace(/^\/+/, "").replace(/\/+$/, "");
}
